infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java

```diff
@@ -0,0 +1,79 @@
+package org.corfudb.infrastructure.log;
+
+import lombok.Getter;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.view.Address;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * A container object that holds log tail offsets and the global
+ * log tail that has been seen. Note that holes don't belong to any
+ * stream therefore the globalTail needs to be tracked separately.
+ *
+ * <p>Created by maithem on 10/15/18.
+ */
+
+@NotThreadSafe
+@ToString
+@Slf4j
+public class LogMetadata {
+
+    @Getter
+    private volatile long globalTail;
+
+    @Getter
+    private final Map<UUID, Long> streamTails;
+
+    public LogMetadata() {
+        this.globalTail = Address.NON_ADDRESS;
+        this.streamTails = new HashMap();
+    }
+
+    public void update(List<LogData> entries) {
+        for (LogData entry : entries) {
+            update(entry);
+        }
+    }
+
+    public void update(LogData entry) {
+        long entryAddress = entry.getGlobalAddress();
+        updateGlobalTail(entryAddress);
+        for (UUID streamId : entry.getStreams()) {
+            long currentStreamTail = streamTails.getOrDefault(streamId, Address.NON_ADDRESS);
+            streamTails.put(streamId, Math.max(currentStreamTail, globalTail));
```

Comment:
This should be the max of currentStreamTail and entryAddress, using the globalTail will result in incorrect tail construction.

Reply:
Done.